instrumentation/servlet/src/main/java/brave/servlet/internal/ServletRuntime.java

```diff
@@ -91,43 +92,54 @@ private static ServletRuntime findServletRuntime() {
     }
 
     @Override public void handleAsync(
-      HttpServerHandler<brave.http.HttpServerRequest, brave.http.HttpServerResponse> handler,
+      HttpServerHandler<HttpServerRequest, HttpServerResponse> handler,
       HttpServletRequest request, HttpServletResponse response, Span span) {
       if (span.isNoop()) return; // don't add overhead when we aren't httpTracing
       TracingAsyncListener listener = new TracingAsyncListener(handler, span);
       request.getAsyncContext().addListener(listener, request, response);
     }
 
     static final class TracingAsyncListener implements AsyncListener {
-      final HttpServerHandler<brave.http.HttpServerRequest, brave.http.HttpServerResponse> handler;
+      final HttpServerHandler<HttpServerRequest, HttpServerResponse> handler;
       final Span span;
-      volatile boolean complete; // multiple async events can occur, only complete once
+      final AtomicBoolean complete = new AtomicBoolean(); // multiple async events can occur
+      final AtomicBoolean timeout = new AtomicBoolean(); // multiple timeout events can occur
 
       TracingAsyncListener(
-        HttpServerHandler<brave.http.HttpServerRequest, brave.http.HttpServerResponse> handler,
+        HttpServerHandler<HttpServerRequest, HttpServerResponse> handler,
         Span span
       ) {
         this.handler = handler;
         this.span = span;
       }
 
       @Override public void onComplete(AsyncEvent e) {
-        if (complete) return;
-        handler.handleSend(httpServerResponse(e), null, span);
-        complete = true;
+        if (!complete.compareAndSet(false, true)) return; // already completed
+        HttpServletRequest req = (HttpServletRequest) e.getSuppliedRequest();
+        HttpServletResponse res = (HttpServletResponse) e.getSuppliedResponse();
+        HttpServerResponse response =
+          brave.servlet.HttpServletResponseWrapper.create(req, res, e.getThrowable());
+        handler.handleSend(response, response.error(), span);
       }
 
+      // Per Servlet 3 section 2.3.3.3, we can't see the final HTTP status, yet. defer to onComplete
+      // https://download.oracle.com/otndocs/jcp/servlet-3.0-mrel-eval-oth-JSpec/
       @Override public void onTimeout(AsyncEvent e) {
-        if (complete) return;
-        span.tag("error", String.format("Timed out after %sms", e.getAsyncContext().getTimeout()));
-        handler.handleSend(httpServerResponse(e), null, span);
-        complete = true;
+        if (!timeout.compareAndSet(false, true)) return; // already timed out
```

Comment:
I noticed timeout hitting many times in the same request, but I might double-check this before merge.

Reply:
we weren't and what we are doing in this callback isn't worth guarding